File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Leave.java
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Scoreboard;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.Command;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.DisplaySlot;

import java.util.List;

public class Leave extends ICommand {

    public Leave(String name) {
        super(name);
    }

    @Override
    public void onPlayerCommand(Player sp, Command cmd, String label, String[] args) {
        if (TosoGameAPI.isBroadCaster(sp) || TosoGameAPI.hasPermission(sp)) {
            if (args.length != 0) {
                if (TosoGameAPI.isBroadCaster(sp) || TosoGameAPI.isAdmin(sp)) {
                    for (String name : args) {
                        Player target = Bukkit.getPlayerExact(name);
                        if (target == null) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + name + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                            continue;
                        } else {
                            Player p = Bukkit.getPlayerExact(name);
                            if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
                                Teams.joinTeam(Teams.OnlineTeam.TOSO_ADMIN, p);
                                p.setGameMode(GameMode.CREATIVE);

                                TosoGameAPI.removeArmor(p);
                                TosoGameAPI.setItem(WorldManager.GameType.START, p);
                                TosoGameAPI.setPotionEffect(p);
                                TosoGameAPI.addOp(p);

                                org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(p);
                                board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);

                                Main.opGamePlayerSet.remove(p.getUniqueId());

                                TosoGameAPI.showPlayers(p);
                                TosoGameAPI.hidePlayers(p);

                                if (MissionManager.isBossBar())
                                    MissionManager.getBossBar().addPlayer(p);

                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + p.getName() + "を運営に追加しました。");
                                p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "あなたを運営に追加しました。");
                                Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "がゲームから離脱しました。");
                                continue;
                            }
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでにゲームから離脱しています。");
                            continue;
                        }
                    }
                    return;
                }
                MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
                return;
            } else {
                if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, sp)) {
                    Teams.joinTeam(Teams.OnlineTeam.TOSO_ADMIN, sp);
                    sp.setGameMode(GameMode.CREATIVE);

                    TosoGameAPI.removeArmor(sp);
                    TosoGameAPI.setItem(WorldManager.GameType.START, sp);
                    TosoGameAPI.setPotionEffect(sp);
                    TosoGameAPI.addOp(sp);

                    org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(sp);
                    board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);

                    Main.opGamePlayerSet.remove(sp.getUniqueId());

                    TosoGameAPI.showPlayers(sp);
                    TosoGameAPI.hidePlayers(sp);

                    if (MissionManager.isBossBar())
                        MissionManager.getBossBar().addPlayer(sp);

                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "あなたを運営に追加しました。");
                    Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + sp.getName() + "がゲームから離脱しました。");
                    return;
                }
                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "すでにゲームから離脱しています。");
                return;
            }
        }
        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
        return;
    }

    @Override
    public void onBlockCommand(BlockCommandSender bs, Command cmd, String label, String[] args) {
        if (args.length != 0) {
            for (String name : args) {
                Player target = Bukkit.getPlayerExact(name);
                if (target == null) {
                    bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + name + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                    continue;
                } else {
                    Player p = Bukkit.getPlayerExact(name);
                    if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
                        Teams.joinTeam(Teams.OnlineTeam.TOSO_ADMIN, p);
                        p.setGameMode(GameMode.CREATIVE);

                        TosoGameAPI.removeArmor(p);
                        TosoGameAPI.setItem(WorldManager.GameType.START, p);
                        TosoGameAPI.setPotionEffect(p);
                        TosoGameAPI.addOp(p);

                        org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(p);
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);

                        Main.opGamePlayerSet.remove(p.getUniqueId());

                        TosoGameAPI.showPlayers(p);
                        TosoGameAPI.hidePlayers(p);

                        if (MissionManager.isBossBar())
                            MissionManager.getBossBar().addPlayer(p);

                        bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + p.getName() + "を運営に追加しました。");
                        p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "あなたを運営に追加しました。");
                        Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "がゲームから離脱しました。");
                        continue;
                    }
                    bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでにゲームから離脱しています。");
                    continue;
                }
            }
            return;
        }
        MainAPI.sendMessage(bs, MainAPI.ErrorMessage.ARGS_PLAYER);
        return;
    }

    @Override
    public void onConsoleCommand(ConsoleCommandSender cs, Command cmd, String label, String[] args) {
        if (args.length != 0) {
            for (String name : args) {
                Player target = Bukkit.getPlayerExact(name);
                if (target == null) {
                    cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + name + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                    continue;
                } else {
                    Player p = Bukkit.getPlayerExact(name);
                    if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
                        Teams.joinTeam(Teams.OnlineTeam.TOSO_ADMIN, p);
                        p.setGameMode(GameMode.CREATIVE);

                        TosoGameAPI.removeArmor(p);
                        TosoGameAPI.setItem(WorldManager.GameType.START, p);
                        TosoGameAPI.setPotionEffect(p);
                        TosoGameAPI.addOp(p);

                        org.bukkit.scoreboard.Scoreboard board = Scoreboard.getBoard(p);
                        board.getObjective(TosoGameAPI.Objective.SIDEBAR.getName()).setDisplaySlot(DisplaySlot.SIDEBAR);

                        Main.opGamePlayerSet.remove(p.getUniqueId());

                        TosoGameAPI.showPlayers(p);
                        TosoGameAPI.hidePlayers(p);

                        if (MissionManager.isBossBar())
                            MissionManager.getBossBar().addPlayer(p);

                        cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + p.getName() + "を運営に追加しました。");
                        p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "あなたを運営に追加しました。");
                        Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "がゲームから離脱しました。");
                        continue;
                    }
                    cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでにゲームから離脱しています。");
                    continue;
                }
            }
            return;
        }
        MainAPI.sendMessage(cs, MainAPI.ErrorMessage.ARGS_PLAYER);
        return;
    }

    @Override
    public List<String> onPlayerTabComplete(Player sp, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onBlockTabComplete(BlockCommandSender bs, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onConsoleTabComplete(ConsoleCommandSender cs, Command cmd, String alias, String[] args) {
        return null;
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Leave.kt
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand
import jp.aoichaan0513.A_TosoGame_Live.Mission.HunterZone
import jp.aoichaan0513.A_TosoGame_Live.Utils.isAdminTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.setTeam
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.Command
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player

class Leave(name: String) : ICommand(name) {

    override fun onPlayerCommand(sp: Player, cmd: Command, label: String, args: Array<String>) {
        if (TosoGameAPI.isBroadCaster(sp) || TosoGameAPI.hasPermission(sp)) {
            if (args.isNotEmpty()) {
                if (TosoGameAPI.isBroadCaster(sp) || TosoGameAPI.isAdmin(sp)) {
                    for (name in args.filter { it.isNotEmpty() }) {
                        val p = Bukkit.getPlayerExact(name)
                        if (p != null) {
                            if (!p.isAdminTeam) {
                                p.setTeam(Teams.OnlineTeam.TOSO_ADMIN, false)
                                HunterZone.removeJoinedSet(p)

                                sp.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}${p.name}を運営に追加しました。")
                                MainAPI.broadcastAdminMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.GREEN}がゲームから抜けました。")
                                continue
                            }
                            sp.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.ERROR)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.RED}はすでにゲームに抜けています。")
                            continue
                        }
                        MainAPI.sendOfflineMessage(sp, name)
                    }
                    return
                }
                MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS)
            } else {
                if (!sp.isAdminTeam) {
                    sp.setTeam(Teams.OnlineTeam.TOSO_ADMIN, false)
                    HunterZone.removeJoinedSet(sp)

                    MainAPI.broadcastAdminMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${sp.name}${ChatColor.RESET}${ChatColor.GREEN}がゲームから抜けました。")
                    return
                }
                sp.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.ERROR)}すでにゲームから離脱しています。")
            }
            return
        }
        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS)
    }

    override fun onBlockCommand(bs: BlockCommandSender, cmd: Command, label: String, args: Array<String>) {
        if (args.isNotEmpty()) {
            for (name in args.filter { it.isNotEmpty() }) {
                val p = Bukkit.getPlayerExact(name)
                if (p != null) {
                    if (!p.isAdminTeam) {
                        p.setTeam(Teams.OnlineTeam.TOSO_ADMIN, false)
                        HunterZone.removeJoinedSet(p)

                        bs.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}${p.name}を運営に追加しました。")
                        MainAPI.broadcastAdminMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.GREEN}がゲームから抜けました。")
                        continue
                    }
                    bs.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.ERROR)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.RED}はすでにゲームに抜けています。")
                    continue
                }
                MainAPI.sendOfflineMessage(bs, name)
            }
            return
        }
        MainAPI.sendMessage(bs, MainAPI.ErrorMessage.ARGS_PLAYER)
    }

    override fun onConsoleCommand(cs: ConsoleCommandSender, cmd: Command, label: String, args: Array<String>) {
        if (args.isNotEmpty()) {
            for (name in args.filter { it.isNotEmpty() }) {
                val p = Bukkit.getPlayerExact(name)
                if (p != null) {
                    if (!p.isAdminTeam) {
                        p.setTeam(Teams.OnlineTeam.TOSO_ADMIN, false)
                        HunterZone.removeJoinedSet(p)

                        cs.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}${p.name}を運営に追加しました。")
                        MainAPI.broadcastAdminMessage("${MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.GREEN}がゲームから抜けました。")
                        continue
                    }
                    cs.sendMessage("${MainAPI.getPrefix(MainAPI.PrefixType.ERROR)}${ChatColor.BOLD}${ChatColor.UNDERLINE}${p.name}${ChatColor.RESET}${ChatColor.RED}はすでにゲームに抜けています。")
                    continue
                }
                MainAPI.sendOfflineMessage(cs, name)
            }
            return
        }
        MainAPI.sendMessage(cs, MainAPI.ErrorMessage.ARGS_PLAYER)
    }

    override fun onPlayerTabComplete(sp: Player, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }

    override fun onBlockTabComplete(bs: BlockCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }

    override fun onConsoleTabComplete(cs: ConsoleCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }
}